File: 02/Day02.hs
module Day02 (main) where

import Flow
import System.IO

data Result = Draw | Lose | Win deriving (Show)
data Shape = Rock | Paper | Scissors deriving (Show)

main :: IO ()
main = do
    guideData <- readFile "./02/input.txt"
    let rounds = guideData |> lines |> map words
    let assumedPoints = rounds
            |> map (\[opp, res] -> (shape opp, shape res))
            |> map (\(oppShape, resShape) -> (shapePoints resShape) + (resultPoints (outcome oppShape resShape)))
            |> sum
    putStrLn $ "Total score (assumed): " ++ (show assumedPoints)
    let directedPoints = rounds
            |> map (\[opp, res] -> (shape opp, targetShape (result res) (shape opp)))
            |> map (\(oppShape, targetShape) -> (shapePoints targetShape) + (resultPoints (outcome oppShape targetShape)))
            |> sum
    putStrLn $ "Total score (directed): " ++ (show directedPoints)

outcome :: Shape -> Shape -> Result
outcome Rock Rock = Draw
outcome Rock Paper = Win
outcome Rock Scissors = Lose
outcome Paper Rock = Lose
outcome Paper Paper = Draw
outcome Paper Scissors = Win
outcome Scissors Rock = Win
outcome Scissors Paper = Lose
outcome Scissors Scissors = Draw

result :: String -> Result
result "X" = Lose
result "Y" = Draw
result "Z" = Win

resultPoints :: Result -> Int
resultPoints Lose = 0
resultPoints Draw = 3
resultPoints Win = 6

shape :: String -> Shape
shape "A" = Rock
shape "X" = Rock
shape "B" = Paper
shape "Y" = Paper
shape "C" = Scissors
shape "Z" = Scissors

shapePoints :: Shape -> Int
shapePoints Rock = 1
shapePoints Paper = 2
shapePoints Scissors = 3

targetShape :: Result -> Shape -> Shape
targetShape Draw shape = shape
targetShape Lose Rock = Scissors
targetShape Lose Paper = Rock
targetShape Lose Scissors = Paper
targetShape Win Rock = Paper
targetShape Win Paper = Scissors
targetShape Win Scissors = Rock

File: 02/day02.ts
import { readLines } from "https://deno.land/std@0.171.0/io/mod.ts";

const fileReader = await Deno.open("./02/input.txt");

enum Result {
  Draw = 3,
  Lose = 0,
  Win = 6,
}

enum Shape {
  Rock = "ROCK",
  Paper = "PAPER",
  Scissors = "SCISSORS",
}

function outcomePoints(opp: Shape, move: Shape): Result {
  if (opp === Shape.Rock) {
    if (move === Shape.Rock) return Result.Draw;
    if (move === Shape.Paper) return Result.Win;
    if (move === Shape.Scissors) return Result.Lose;
  }
  if (opp === Shape.Paper) {
    if (move === Shape.Rock) return Result.Lose;
    if (move === Shape.Paper) return Result.Draw;
    if (move === Shape.Scissors) return Result.Win;
  }
  if (opp === Shape.Scissors) {
    if (move === Shape.Rock) return Result.Win;
    if (move === Shape.Paper) return Result.Lose;
    if (move === Shape.Scissors) return Result.Draw;
  }
  throw new Error("Invalid moves");
}

function shapePoints(shape: Shape): number {
  switch (shape) {
    case Shape.Rock:
      return 1;
    case Shape.Paper:
      return 2;
    case Shape.Scissors:
      return 3;
    default:
      throw new Error("Invalid shape");
  }
}

function targetShape(result: Result, shape: Shape): Shape {
  switch (result) {
    case Result.Draw:
      return shape;
    case Result.Lose:
      if (shape === Shape.Rock) return Shape.Scissors;
      if (shape === Shape.Paper) return Shape.Rock;
      if (shape === Shape.Scissors) return Shape.Paper;
      break;
    case Result.Win:
      if (shape === Shape.Rock) return Shape.Paper;
      if (shape === Shape.Paper) return Shape.Scissors;
      if (shape === Shape.Scissors) return Shape.Rock;
      break;
  }
  throw new Error("Invalid win shapes");
}

function toResult(code: string): Result {
  switch (code) {
    case "X":
      return Result.Lose;
    case "Y":
      return Result.Draw;
    case "Z":
      return Result.Win;
  }
  throw new Error("Invalid result code");
}

function toShape(code: string): Shape {
  switch (code) {
    case "A":
    case "X":
      return Shape.Rock;
    case "B":
    case "Y":
      return Shape.Paper;
    case "C":
    case "Z":
      return Shape.Scissors;
  }
  throw new Error("Invalid shape code");
}

let score = 0;
let adjustedScore = 0;

for await (let round of readLines(fileReader)) {
  const [oppString, code] = round.split(" ");
  const opp = toShape(oppString);

  // Part One
  const promptedMove = toShape(code);
  score += shapePoints(promptedMove) + outcomePoints(opp, promptedMove);

  // Part Two
  const result = toResult(code);
  const nextMove = targetShape(result, opp);
  adjustedScore += shapePoints(nextMove) + outcomePoints(opp, nextMove);
}

console.log("Total score:", score);
console.log("Adjusted total score:", adjustedScore);
